crates/compiler/mono/src/inc_dec.rs

```diff
@@ -1,1554 +1,1193 @@
+                        "The current consumption should be a superset of the previous consumption.
+                        As the consumption should only ever increase.
+                        Otherwise we will be looping forever."
+                    );
+                    joinpoint_consumption = current_joinpoint_consumption;
+                }
+            };
 
-                self.arena.alloc(Stmt::Let(z, v, l, b))
-            }
+            environment.add_joinpoint_consumption(*joinpoint_id, joinpoint_consumption);
+            let new_remainder = insert_refcount_operations_stmt(arena, environment, remainder);
+            environment.remove_joinpoint_consumption(*joinpoint_id);
 
-            EmptyArray | Literal(_) | Reset { .. } | NullPointer | RuntimeErrorFunction(_) => {
-                // EmptyArray is always stack-allocated function pointers are persistent
-                self.arena.alloc(Stmt::Let(z, v, l, b))
+            arena.alloc(Stmt::Join {
+                id: *joinpoint_id,
+                parameters,
+                body: new_body,
+                remainder: new_remainder,
+            })
+        }
+        Stmt::Jump(joinpoint_id, arguments) => {
+            let consumed_symbols = environment.get_joinpoint_consumption(*joinpoint_id);
+            for consumed_symbol in consumed_symbols.iter() {
+                environment.consume_symbol(consumed_symbol);
             }
-        };
-
-        (new_b, live_vars)
-    }
-
-    fn update_var_info(&self, symbol: Symbol, layout: &InLayout<'a>, expr: &Expr<'a>) -> Self {
-        // is this value a constant? TODO do function pointers also fall into this category?
-        let persistent = false;
 
-        // must this value be consumed?
-        let consume = consume_expr(&self.vars, expr);
+            let new_jump = arena.alloc(Stmt::Jump(*joinpoint_id, arguments));
 
-        let reset = matches!(expr, Expr::Reset { .. });
-
-        self.update_var_info_help(symbol, layout, persistent, consume, reset)
+            // Note that this should only insert increments if a later join point has a current parameter as consumed closure.
+            consume_and_insert_inc_stmts(
+                arena,
+                environment,
+                environment.owned_usages(arguments.iter().copied()),
+                new_jump,
+            )
+        }
+        Stmt::Crash(symbol, crash_tag) => {
+            // We don't have to worry about reference counting *after* the crash.
+            // But we do need to make sure the symbol of the crash is live until the crash.
+            // So we insert increment statements for the symbol (if it is reference counted)
+            let new_crash = arena.alloc(Stmt::Crash(*symbol, *crash_tag));
+
+            consume_and_insert_inc_stmts(
+                arena,
+                environment,
+                environment.owned_usages(std::iter::once(symbol).copied()),
```

Comment:
you could use a 1-element array as the iterator here: `[*symbol]`

Reply:
What is the benefit of that approach?

Reply:
it has fewer characters. that's it really